### PR TITLE
Support Custom JTAG Cycle Protocols and Clean Up

### PR DESCRIPTION
**Pull request by ginty**
This PR includes the following:

* Adds a new feature to allow applications to define how JTAG cycles are implemented at the pin-level.
* Updated all documentation and internal code to reference TCK instead of TCLK, though in all cases the TCLK versions of pin names and options are still supported.
* Adds documentation for the new feature (below) and some general clean up of the existing docs - moves the legacy integration description to another page to give it less prominence going forward.

#### Custom JTAG Cycle Protocols

This plugin also allows applications to define how a JTAG cycle should be implemented at the
pin-level, providing the flexibility to handle hybrid JTAG protocols.
An example would be an ultra-low pincount device where the JTAG signals are multiplexed down a
single wire.

To implement a custom protocol, a `:cycle_callback` option should be supplied when instantiating the
driver to specify the name of the DUT method that will implement the cycle.
**Note that most other configuration options are meaningless in this context (and some non-default
values will raise an error) because the application
is now taking full responsibility for how a JTAG cycle is implemented.**
Similarly, there is no requirement for the conventional JTAG pins to be present or specified when this
option is given since the driver is no longer responsible for the vector generation.

The given method should accept two arguments:

* **actions** - A hash containing the actions that should be applied to the `:tdi`, `:tms` and `:tdo`,
  pins and a boolean flag called `:store` which indicates if TDO should be stored/captured on this cycle
* **options** - Any options that should be passed to the tester when calling `tester.cycle`

The JTAG driver provides a method called `apply_action` which accepts a pin object and the pin-specific
actions from the actions hash, e.g.

~~~ruby
jtag.apply_action(pin(:my_pin), actions[:tms])
~~~

Here is a complete example which would serialize TDI, TMS and TDO on a single pin:

~~~ruby
class MyApp::MyDUT
  def initialize(options = {})
    add_pin :tck
    add_pin :tio

    sub_block :jtag, class_name: 'OrigenJTAG::Driver', cycle_callback: :jtag_cycle
  end

  def jtag_cycle(actions, options = {})
    pin(:tck).drive(1)
    jtag.apply_action(pin(:tio), actions[:tdi])
    tester.cycle(options)
    jtag.apply_action(pin(:tio), actions[:tms])
    tester.cycle(options)
    jtag.apply_action(pin(:tio), actions[:tdo])
    tester.store_next_cycle(pin(:tio)) if actions[:store]
    tester.cycle(options)
  end
end
~~~

### CODE REVIEW

**Comment by ginty** [comment]
@pderouen, there are a lot of changes that show up here in the driver, but the overall architecture is the same and most of it is variable renaming cleanup and replacing pin operations like this:

~~~ruby
@tdi_pin.drive(1)
~~~

with this instead (which just records the intent for later application instead of applying it straight away):

~~~ruby
action :tdi, :drive, 1
~~~

However, I think the final solution I ended up with here could enable significant simplification of the driver - there is a lot of complexity right now that is quite hard to reason about to handle the various multi-cycle options that it currently supports.
Specifically, I think the `tclk_cycle` method could be removed completely and the `shift` method could be simplified considerably - with the new system I have in mind, the `shift` method would only need to worry about the regular single-cycle case.

In this update I have added a `cycle` method which handles application of data states to the pins:

~~~ruby
def cycle(options = {})
  if @actions
    if cycle_callback?
      @owner.send(@cycle_callback, @actions, options)
    else
      apply_action(@pins[:tms], @actions[:tms])
      apply_action(@pins[:tdi], @actions[:tdi])
      apply_action(@pins[:tdo], @actions[:tdo])
      tester.store_next_cycle(@pins[:tdo]) if @actions[:store]
      tester.cycle(options)
    end
    clear_actions
  end
end
~~~

So it basically says either handover to the application to update the pins, or else apply the current actions to the conventional JTAG pins.

However, I think pretty much all of the multi-cycle options could be fully encapsulated within this method. So the rest of the driver would be concerned with generating the sequence to be applied to the JTAG signals, and then the cycle method would deal with how they are implemented at pin-level based on the current configuration.

For example, to make it handle a vector-driven TCK would be easy:

~~~ruby
apply_action(@pins[:tms], @actions[:tms])
apply_action(@pins[:tdi], @actions[:tdi])
apply_action(@pins[:tdo], @actions[:tdo])
if tck_multiple == 1
  @pins[:tck].drive(1)
  tester.store_next_cycle(@pins[:tdo]) if @actions[:store]
  tester.cycle(options)
elsif tck_multiple == 2
  @pins[:tck].drive(1)
  tester.cycle(options)
  @pins[:tck].drive(0)
  tester.store_next_cycle(@pins[:tdo]) if @actions[:store]
  tester.cycle(options)
end
~~~

Obviously the final implementation would have a lot more logic, and not saying that the final method wouldn't be complex, but it would all be in one place and should be pretty easy to follow.

What do you think?

I am happy to make the change btw, but I didn't want to embark on a significant refactor like that without consulting with you first.

Also, I noticed while implementing this change that this line can be commented out completely (and is in this PR) without any affect on the tests: https://github.com/Origen-SDK/origen_jtag/blob/master/lib/origen_jtag/driver.rb#L290

Is that OK to remove, or do think it really does something not covered by the tests?

Anyway, I think that supports the fact that some of the internals here are quite hard to follow because it couples the JTAG protocol generation with the cycle-level application of it. My proposal would separate these and I think it would make the overall system easier to understand.

**Comment by pderouen** [comment]
@ginty my initial thoughts:

I've always felt that the JTAG driver code is too complex and too difficult to follow. I wrote the SPI driver the way I would've wanted the JTAG driver to be. I think it's just a side effect of the length of time it's been around. I'm in favor of simplifying it. We have a wealth of test cases that should help ensure backwards compatibility.

I'm in favor of calling it TCK, that's what the standard calls it.

As far as the line that isn't needed goes, there are tests for the capture feature. I did add new tests for the capture feature when I moved the capture/source implementation responsibility to the testers plugin. So, if those tests still pass the line indeed isn't needed.

**Comment by ginty** [comment]
Hi @pderouen, thanks for the timely review.

I chickened out of the discussed refactoring to simplify the internals - although I think it is still valid, it will take a while to do it and its debatable what the value-add is right now.

So, shelving that work for another day...
